src/lang/typing.ml

```diff
@@ -165,97 +169,183 @@ let rec occur_check a b =
+                          else b.descr <- EVar (j, InternalMedia :: c)
+                      | _ -> raise (Unsatisfied_constraint (InternalMedia, b)))
+                | Num -> (
+                    match (demeth b).descr with
+                      | Ground g ->
+                          if g <> Int && g <> Float then
+                            raise (Unsatisfied_constraint (Num, b))
+                      | EVar (j, c) ->
+                          if List.mem Num c then ()
+                          else b.descr <- EVar (j, Num :: c)
+                      | _ -> raise (Unsatisfied_constraint (Num, b))))
+              constraints;
             constraints
-      | _ -> assert false (* only EVars are bindable *)
-    end;
+        | _ -> assert false
+      (* only EVars are bindable *)
+    in
+
+    (* We do not want to check the constraints when we increase the types (for
+       now). *)
+    let variance = if constraints <> [] then Invariant else variance in
 
     (* This is a shaky hack...
      * When a value is passed to a FFI, its type is bound to a type without
      * any location.
      * If it doesn't break sharing, we set the parsing position of
      * that variable occurrence to the position of the inferred type. *)
     if b.pos = None && match b.descr with EVar _ -> false | _ -> true then
-      a.descr <- Link { a0 with descr = b.descr }
-    else a.descr <- Link b)
+      a.descr <- Link (variance, { a0 with descr = b.descr })
+    else a.descr <- Link (variance, b))
 
 (** {1 Subtype checking/inference} *)
 
+exception Incompatible
+
+(** Approximated supremum of two types. We grow the second argument so that it
+    has a chance be be greater than the first. No binding is performed by this
+    function so that it should always be followed by a subtyping. *)
+let rec sup ~pos a b =
+  (* Printf.printf "sup %s // %s\n%!" (Type.print a) (Type.print b); *)
+  let sup = sup ~pos in
+  let mk descr = { level = -1; pos; descr } in
+  let scheme_sup t t' =
+    match (t, t') with ([], t), ([], t') -> ([], sup t t') | _ -> t'
+  in
+  let rec meth_type l a =
+    match (deref a).descr with
+      | Meth (l', t, _, _) when l = l' -> Some t
+      | Meth (_, _, _, a) -> meth_type l a
+      | EVar _ -> Some ([], fresh_evar ~level:(-1) ~pos)
+      | _ -> None
+  in
+  match ((deref a).descr, (deref b).descr) with
+    | EVar _, _ -> b
+    | _, EVar _ -> a
+    | Nullable a, Nullable b -> mk (Nullable (sup a b))
+    | Nullable a, _ -> mk (Nullable (sup a b))
+    | _, Nullable b -> mk (Nullable (sup a b))
+    | List a, List b -> mk (List (sup a b))
+    | Arrow (p, a), Arrow (q, b) ->
+        if List.length p <> List.length q then raise Incompatible;
+        mk (Arrow (q, sup a b))
+    | Tuple l, Tuple m ->
+        if List.length l <> List.length m then raise Incompatible;
+        mk (Tuple (List.map2 sup l m))
+    | Ground g, Ground g' ->
+        (* It might happen that we compare abstract values. *)
```

Comment:
Indeed, this can happen with the extensive abstract type thanks for thinking of that.

Reply:
I didn't think of it, it actually bite me ;)